Approach.md

```diff
@@ -0,0 +1,102 @@
+There will be no network devices attached to any tasks.
+
+Mounting is also considered out of scope and as such each task will be chrooted to a RW root that's a copy of a base image that's removed upon exit of the task.
+
+I'll make no attempt at Linux backwards compatibility and the minimum kernel version will be 5.0+.
+
+Tasks will inherit the UID/GID of the daemon and internally be mapped as root.
+
+For the gRPC implementation I'll be using the rust crate [tonic](https://github.com/hyperium/tonic) to avoid reinventing the wheel.
+
+Mutable interaction with tasks will be mutually exclusive meaning that the stop action will take a write-only lock on the task such that any concurrent stop task will wait for the first to complete and the subsequently fail as the task has already stopped.
+Read-only interactions such as query and stream can happen concurrently without issue.
+
+## rrocker-cli:
+The CLI will need to be run once per command, so scheduling multiple tasks requires multiple invocations.
+
+There'll be zero command line switches and as such it'll be hardcoded to connect to `rrockerd` running on localhost.
+
+## rrocker-lib:
+No attention will be paid to backwards compatibility of the API, meaning no versioning or abstractions.
+
+Human friendliness of the API is a distant afterthought, as such full UUIDs will be required to interact with tasks.
+
+Resource limits will be exposed 1-to-1 with the underlying cgroups API leaving it up to the `rrocker-cli` to expose it in a human friendly way.
+# Security
+Security is naturally an important aspect of the implementation however as my experience with crypto is limited I'll attempt to follow best practices and use the default configuration of OpenSSL and rustTLS wherever applicable.
+In an actual production system you'd have a security expert decide the configuration based on a threat model.
+
+## Authentication
+Authentication will be done with mTLS according to the challenge rules.
+
+The following certificates will be created:
+- Root Certificate Authority (CA). This is our root of trust and should be stored on an air-gapped system that's only used to sign/revoke the server/client CAs.
+- Separate server and client CAs. These are used to sign/revoke server and client certs. Furthermore they're separated such that an infrastructure team is able to own their CA and deploy new servers independently. These CAs also introduce an indirection such that servers and clients don't need to know about all clients and servers but can simply verify an identity has been signed by the server/client CA.
+- Server_1 cert used by the server to auth with the clients
+- Client_1 and Client_2 certs used to demo user's only being able to see their own tasks
+- Admin_1 cert used to demo admin's being able to see all tasks
+- An untrusted CA + client cert to demo unauthorized client's can't connect
```

Comment:
To demo an unauthorized client may not need a CA and can just be self signed... if that helps simplify it at all. But nothing wrong with a separate CA, just wanted to mention it as an option.

Reply:
Good point, no need to introduce an extra CA. I'll fix this.